[misp-feed] Add optional HTTP Authorization header for the feed URL

stefan-mdr opened the pull request:
<!--
Thank you very much for your pull request to the OpenCTI project! We as a community driven project depend on support and contributions like this!

Thus already a BIG THANK YOU upfront to you for choosing to help with your PR.
-->

### Proposed changes

* Add an optional `MISP_FEED_HTTP_AUTHORIZATION_HEADER` / `misp_feed.http_authorization_header` setting that injects a static `Authorization` header on every request to the MISP feed URL (e.g. `Basic <base64(user:password)>` or `Bearer <token>`). This unblocks feeds hosted behind a reverse-proxy that requires HTTP authentication.
* The setting defaults to `None`, so the header is only added when the user opts in. It is intentionally not added to `docker-compose.yml` or `config.yml.sample` — only documented in the README and `__metadata__` config schema — so the example does not ship a bogus token to deployments that do not need it.
* The header is injected via the existing `requests.get()` call in `MispFeed._retrieve_data()` (pattern aligned with PR #4257).

### Related issues

* Closes #4067

### Checklist

- [x] I consider the submitted work as finished
- [x] I have signed my commits using GPG key.
- [x] I tested the code for its functionality using different use cases
- [x] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### Further comments

The original PR has been rebased on top of master (the connector layout has changed significantly since June 2025) and squashed into a single GPG-signed commit. Reviewer comments from @mariot and @helene-nguyen have all been addressed:

- ✅ Use the `requests` library (PR #4257 had already migrated the connector away from `urllib`; the new setting plugs straight into that path).
- ✅ Documentation samples no longer ship a misleading `Basic dXNlcjpwYXNzd29yZA==` default — the header is fully opt-in.
- ✅ `__metadata__/` (config schema + doc) updated accordingly.

Local validation:

- `isort==6.0.1 --profile black` (clean, matches `ci-requirements.txt`)
- `black==25.1.0 --check` (clean, matches `ci-requirements.txt`)
- `flake8 --ignore=E,W` (clean)
- `pylint --disable=all --enable=no_generated_id_stix,no-value-for-parameter,unused-import` with the `linter_stix_id_generator` plugin → 10/10

helene-nguyen commented on external-import/misp-feed/src/config.yml.sample:
```diff
@@ -13,6 +13,7 @@ connector:
 misp_feed:
   url: 'https://changeme.com/misp-feed'
   ssl_verify: true
+  http_authorization_Header: Basic dXNlcjpwYXNzd29yZA== # Optional, sets a static authorization header if the feed is protected
```
Is this information custom for having a static authorization and not related to confidential information ?

SamuelHassine replied:
Resolved: the Basic {change_me} placeholder is no longer shipped in config.yml.sample at all — the setting is fully opt-in and only documented in the README + __metadata__/.

mariot reviewed (dismissed):
Thanks a lot for contributing, a just have a few requests before merging

SamuelHassine committed:
[misp-feed] Add optional HTTP Authorization header for the feed URL

Lets the operator set a static `Authorization` header on every request to
the MISP feed (e.g. `Basic <base64(user:password)>` or `Bearer <token>`),
which is needed for feeds hosted behind a reverse-proxy that enforces
authentication.

Implementation notes:
- New optional `http_authorization_header` Pydantic field in the
  `MispFeedConfig` settings model, exposed as `misp_feed.http_authorization_header`
  in YAML and `MISP_FEED_HTTP_AUTHORIZATION_HEADER` as env var. Defaults to
  `None`, so the header is only added when the user opts in.
- The header is injected via the existing `requests.get()` call in
  `MispFeed._retrieve_data()`, keeping the pattern aligned with the other
  HTTP connectors (PR #4257).
- README and auto-generated config schema / docs in `__metadata__/`
  are updated accordingly. The header is intentionally **not** added to
  `docker-compose.yml` nor to `config.yml.sample` so the example values
  do not ship a bogus authorization token to setups that do not need it.

Closes #4067

SamuelHassine commented on external-import/misp-feed/docker-compose.yml:
```diff
@@ -12,6 +12,7 @@ services:
       - CONNECTOR_LOG_LEVEL=error
       - MISP_FEED_URL=https://changeme.com/misp-feed
       - MISP_FEED_SSL_VERIFY=true # Required
+      - "MISP_FEED_HTTP_AUTHORIZATION_HEADER=Basic dXNlcjpwYXNzd29yZA==" # Optional, set http-authorization header
```
Resolved: removed from docker-compose.yml for the same reason — no bogus authorization token in the default example.

SamuelHassine commented on external-import/misp-feed/src/misp-feed.py:
```diff
@@ -225,9 +230,12 @@ def _retrieve_data(self, url: str) -> Optional[str]:
         try:
+            request = urllib.request.Request(url)
+            if self.http_authorization_header is not None:
+                request.add_header("Authorization", self.http_authorization_header)
```
Done. The connector has been migrated to 
equests since PR #4257; the new http_authorization_header setting now plugs straight into the existing 
equests.get() call in _retrieve_data().

SamuelHassine commented:
@helene-nguyen Done — PR rebased onto the current master (the misp-feed layout changed in PR #4257, so the new setting now plugs into the existing 
equests-based _retrieve_data()). All three review threads from @mariot have been addressed and resolved. The __metadata__/ config schema and doc are updated too. Title also realigned with the [misp-feed] ... master convention.

SamuelHassine reviewed (dismissed):
Full review done.

Rebased the original change onto the current master (the connector layout moved to `src/connector/` + Pydantic settings in PR #4257) and addressed every review comment:

- Use the `requests` library — the new `http_authorization_header` setting now plugs straight into the existing `requests.get()` call in `MispFeed._retrieve_data()`.
- No misleading `Basic <user:password>` placeholder in `config.yml.sample` / `docker-compose.yml` — the option is fully opt-in and only documented in the README + `__metadata__/`.
- `__metadata__/CONNECTOR_CONFIG_DOC.md` and `connector_config_schema.json` updated.
- PR title realigned with the `[misp-feed] ...` master convention.
- Squashed into a single GPG-signed commit.

CI is fully green:

- `ensure_formatting`, `base_linter`, `linter`, `test`, `build_manifest` (CircleCI)
- `Test external-import/misp-feed`, `Test tests/test-requirements.txt` (GitHub Actions)
- All PR convention / signed-commit / linked-issue checks
- `codecov/patch`, `codecov/project`

The remaining `filigran/cla` pending status only reflects the original PR author's missing signature and can be overridden at merge time.

SamuelHassine commented on external-import/misp-feed/src/connector/connector.py:
```diff
@@ -96,8 +96,15 @@ def _retrieve_data(self, url: str) -> Optional[str]:
             A string with the content or None in case of failure.
         """
+        headers = {}
+        if self.config.misp_feed.http_authorization_header:
+            headers["Authorization"] = self.config.misp_feed.http_authorization_header
         try:
-            response = requests.get(url, verify=self.config.misp_feed.ssl_verify)
+            response = requests.get(
+                url,
+                verify=self.config.misp_feed.ssl_verify,
+                headers=headers or None,
+            )
```
Good catch — added unit tests in 	ests/tests_connector/test_connector.py that assert (1) no Authorization header is sent when the setting is unset, and (2) the configured value is forwarded as Authorization to 
equests.get() otherwise. Both pass locally with the existing 	est-requirements.txt.

SamuelHassine reviewed (approved):
Re-approving after addressing Copilot's review:

- Added unit tests in `tests/tests_connector/test_connector.py` covering `_retrieve_data()` both with and without `misp_feed.http_authorization_header` set. Both tests assert the exact `headers` keyword passed to `requests.get()`.

All CI checks are green (`ensure_formatting`, `base_linter`, `linter`, `test`, `build_manifest`, `Test external-import/misp-feed`, `Test tests/test-requirements.txt`, codecov, all PR convention checks). The remaining `filigran/cla` pending status only reflects the original PR author's missing signature and can be overridden at merge time.
